### PR TITLE
storage: improve reliability of SSH tunnels with large clusters

### PR DESCRIPTION
**Pull request by benesch**
*Part of #18490.*

Rather than having separate APIs for rewrites that need to keep a token alive and rewrites that don't, unify the API around closures. A rewrite is now a closure that returns the rewritten URL. This allows the rewrites that need to keep a token alive to do so by embedding the token in the closure, with no special support required by the BrokerRewritingClientContext.

A future commit will use this support to allow the rewrite addresses to change over time to reflect a failed and restarted SSH tunnel.

### Motivation

  * This PR will fix a previously unreported bug: Materialize creates SSH tunnels _per worker_, which results in O(workers * brokers) SSH connections. On larger cluster sizes, this means that hundreds of SSH tunnels are attempted at once, which trips the default `MaxStartups` limit for sshd.

### Tips for reviewer

WIP. More commits coming soon. Just wanted a CI run.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Improve the reliability of SSH tunnels when used with larger cluster sizes.

### CODE REVIEW

**Comment by benesch** [comment]
Noting a comment from Slack for posterity: I deployed this to my staging account, and it worked like a charm. xlarge cluster is happily ingesting data from a Kafka broker via one single SSH tunnel.

**Comment by benesch** [comment]
This is ready for review. I expect it to pass all tests. I'm going to leave the merge conflicts for now, as resolving those merge conflicts would prevent this PR from backporting cleanly into v0.48.x and v0.49.x.

Steps once this is approved:

  * [ ] Backport into v0.48.x, for immediate release into the affected prospect's environment.
  * [ ] Backport into v0.49.x.
  * [ ] Resolve merge conflicts and merge into main.

File: src/ccsr/src/config.rs [hunk 1]
```diff
@@ -76,11 +79,9 @@ impl ClientConfig {
         self
     }
 
-    /// Overrides the url.
-    ///
-    /// Useful for certain kinds of tunneling.
-    pub fn override_url(mut self, url: Url) -> ClientConfig {
-        self.url = url;
+    /// Adds a `Proxy` to the list of proxies that the client will use.
+    pub fn proxy(mut self, proxy: Proxy) -> ClientConfig {
```
Comment (aljoscha) [inline src/ccsr/src/config.rs:83]: very nit: maybe call it `add_proxy` (previously, the could be only one override, but now we can have multiple proxies)

File: test/ssh-connection/kafka-source.td [hunk 2]
```diff
@@ -14,10 +14,19 @@ $ kafka-create-topic topic=thetopic
 $ kafka-ingest topic=thetopic format=bytes
 one
 
+# Create a dedicated cluster for the source, so we can easily restart the
+# source by dropping and recreating the clsuter's replica.
```
Comment (aljoscha) [inline test/ssh-connection/kafka-source.td:18]: typo: clsuter

**Review by aljoscha — APPROVED** [review]
I'm not Gus, but I reviewed it anyways: this looks very nice! Seems reasonable to wait for Gus' opinion, though.

File: src/storage-client/src/ssh_tunnels.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+        loop {
+            // NOTE: this code is structured awkwardly to convince rustc that
+            // the lock is not held across an await point. rustc's analysis
+            // does not take into account explicit `drop` calls, so we have to
+            // structure such that the lock guard goes out of scope.
+            // See: https://github.com/rust-lang/rust/issues/69663
+            enum Action {
+                Return(ManagedSshTunnelHandle),
+                AwaitConnection(future::Shared<oneshot::Receiver<()>>),
```
Comment (guswynn) [inline src/storage-client/src/ssh_tunnels.rs:78]: nit: you might prefer: https://docs.rs/tokio/latest/tokio/sync/watch/index.html here

File: src/storage-client/src/ssh_tunnels.rs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    fn drop(&mut self) {
+        let mut tunnels = self.manager.tunnels.lock().expect("lock poisoned");
+        // If there are only two strong references, the manager holds one and we
+        // hold the other, so this is the last handle.
+        //
+        // IMPORTANT: We must be holding the lock when we perform this check, to
+        // ensure no other threads can acquire a new handle via the manager.
+        if Arc::strong_count(&self.handle) == 2 {
+            tunnels.remove(&self.key);
+        }
```
Comment (guswynn) [inline src/storage-client/src/ssh_tunnels.rs:231]: very nice! I was gonna suggest a `Weak`, but I think this is better!

**Review by guswynn — APPROVED** [review]
looks good!

**Comment by guswynn** [comment]
No idea how this ever passed ci without:
https://github.com/MaterializeInc/materialize/compare/837277c2557eeb493883dbf1b4d4cc41e1f31f40..c83b38d39b1f4326a6ac5e56c4cce982ea3e2f88

**Comment by benesch** [comment]
Hang on, something sees sketchy with that commit @guswynn! Double check me on this, but `kill` doesn't remove the container. It just stops it. The public key should still be there.

**Comment by guswynn** [comment]
@benesch I checked locally and `/etc/authorized_keys` is empty when we restart the bastion

**Comment by guswynn** [comment]
@benesch could this new line:
```
"volumes": [f"{setup_path}:/etc/entrypoint.d/setup.sh"],
```

be altering the behavior?

**Pull request merged**